File: UserProject/src/main/java/com/botthebuilder/userproject/entities/Project.java
package com.botthebuilder.userproject.entities;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.util.List;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Project {

    @Id
    @SequenceGenerator(
            name = "project_sequence",
            sequenceName = "project_sequence",
            allocationSize = 1
    )
    @GeneratedValue(
            strategy = GenerationType.SEQUENCE,
            generator = "project_sequence"
    )
    private Long projectId;

    private String projectName;

    private String botName;

    private String description;

    private String avatarUrl;

    private String api;

    private String apiKey;

    private String modelPath;

    private String dataFilePath;

    private Integer state;

    private String date;

    private Integer totalRequests = 0;

    private Integer totalRatings = 0;

    private Integer sumRatings = 0;

    @ManyToOne(
            optional = false
    )
    @JoinColumn(
            name= "userId",
            referencedColumnName = "userId"
    )
    private User user;

    @OneToMany(
            mappedBy = "project",
            cascade = CascadeType.ALL
    )
    private List<Action> actions;

}

File: PythonMainServer/models/Project.py
from db import db 
from models.Action import Action
from utils.SecretKeyGenerator import Generator

class Project(db.Model):
	project_id = db.Column(db.BigInteger, primary_key = True)
	project_name = db.Column(db.String(20), nullable = False)
	bot_name = db.Column(db.String(20))
	description = db.Column(db.String(20))
	avatar_url = db.Column(db.String(80))
	api_key = db.Column(db.String(80))
	state = db.Column(db.Integer)
	total_requests = db.Column(db.Integer)
	total_ratings = db.Column(db.Integer)
	sum_ratings = db.Column(db.Integer)
	user_id = db.Column(db.BigInteger)
	actions = db.relationship('Action', backref = 'project', lazy = True)

	@classmethod
	def findById(cls, project_id):
		return Project.query.get(project_id)

	def incrementRequests(self):
		if self.total_requests is None:
			self.total_requests = 1
		else:
			self.total_requests += 1
		db.session.commit()

	def setSecretKey(self):
		self.api_key = Generator.generate()
		db.session.commit()


